libtenzir/builtins/operators/from_to_2.cpp

```diff
@@ -582,9 +645,702 @@ class to_plugin2 final : public virtual operator_factory_plugin {
+              .note("while reading `{}`", path)
+              .emit(*dh_);
+            return;
+          }
+          if (args_.remove) {
+            // There is no async call available.
+            auto status = fs_->DeleteFile(path);
+            if (not status.ok()) {
+              diagnostic::warning("failed to remove `{}`", path)
+                .primary(args_.url)
+                .note(status.ToStringWithoutContextLines())
+                .emit(*dh_);
+            }
+          }
+          check_jobs_and_termination();
+        });
+        self_->mail(atom::start_v)
+          .request(executor, caf::infinite)
+          .then([] {},
+                [this, path](caf::error error) {
+                  diagnostic::error(std::move(error))
+                    .note("while reading `{}`", path)
+                    .emit(*dh_);
+                });
+      });
+  }
+
+  auto register_metrics(uint64_t nested_operator_index, uuid nested_metrics_id,
+                        type schema) -> caf::result<void> {
+    (void)nested_operator_index;
+    return self_->mail(operator_index_, nested_metrics_id, std::move(schema))
+      .delegate(metrics_receiver_);
+  }
+
+  auto handle_metrics(uint64_t nested_operator_index, uuid nested_metrics_id,
+                      record metrics) -> caf::result<void> {
+    (void)nested_operator_index;
+    return self_->mail(operator_index_, nested_metrics_id, std::move(metrics))
+      .delegate(metrics_receiver_);
+  }
+
+  from_file_actor::pointer self_;
+  std::unique_ptr<diagnostic_handler> dh_;
+  std::shared_ptr<arrow::fs::FileSystem> fs_;
+
+  // The configuration and things derived from it.
+  from_file_args args_;
+  event_order order_;
+  glob glob_;
+  std::string root_path_;
+
+  // Watching is implemented by checking against the files seen previously.
+  file_set previous_;
+  file_set current_;
+
+  // Communication with the operator bridges.
+  std::deque<caf::typed_response_promise<table_slice>> gets_;
+  std::deque<std::pair<table_slice, caf::typed_response_promise<void>>> puts_;
+
+  // Information needed for spawning subpipelines.
+  std::string definition_;
+  node_actor node_;
+  bool is_hidden_;
+
+  // Job management.
+  size_t active_jobs_ = 0;
+  std::deque<arrow::fs::FileInfo> jobs_;
+  bool added_all_jobs_ = false;
+
+  // Forwarding metrics.
+  uint64_t operator_index_ = 0;
+  metrics_receiver_actor metrics_receiver_;
+};
+
+class from_file final : public crtp_operator<from_file> {
+public:
+  from_file() = default;
+
+  explicit from_file(from_file_args args) : args_{std::move(args)} {
+  }
+
+  auto name() const -> std::string override {
+    return "from_file";
+  }
+
+  auto location() const -> operator_location override {
+    return operator_location::local;
+  }
+
+  auto operator()(operator_control_plane& ctrl) const
+    -> generator<table_slice> {
+    // Spawning the actor detached because some parts of the Arrow filesystem
+    // API are blocking.
```

Comment:
As far as I can tell it's only delete/write? Write returns instantly for every filesystem that Arrow supports because it's effectively just `memcpy` followed by an async flush that does the network request. I'm not sure about deletion.

Reply:
We're resolving this for now.